patroni/ha.py

```diff
@@ -922,18 +930,29 @@ def _is_healthiest_node(self, members: Collection[Member], check_replication_lag
         # Prepare list of nodes to run check against
         members = [m for m in members if m.name != self.state_handler.name and not m.nofailover and m.api_url]
 
-        for st in self.fetch_nodes_statuses(members):
-            if st.failover_limitation() is None:
-                if st.in_recovery is False:
-                    logger.warning('Primary (%s) is still alive', st.member.name)
-                    return False
-                if my_wal_position < st.wal_position:
-                    logger.info('Wal position of %s is ahead of my wal position', st.member.name)
-                    # In synchronous mode the former leader might be still accessible and even be ahead of us.
-                    # We should not disqualify himself from the leader race in such a situation.
-                    if not self.sync_mode_is_active() or not self.cluster.sync.leader_matches(st.member.name):
+        if members:
+            for st in self.fetch_nodes_statuses(members):
+                if st.failover_limitation() is None:
+                    if st.in_recovery is False:
+                        logger.warning('Primary (%s) is still alive', st.member.name)
+                        return False
+                    if my_wal_position < st.wal_position:
+                        logger.info('Wal position of %s is ahead of my wal position', st.member.name)
+                        # In synchronous mode the former leader might be still accessible and even be ahead of us.
+                        # We should not disqualify himself from the leader race in such a situation.
+                        if not self.is_synchronous_mode() or self.cluster.sync.is_empty\
+                                or not self.cluster.sync.leader_matches(st.member.name):
+                            return False
+                        logger.info('Ignoring the former leader being ahead of us')
+                    if my_wal_position == st.wal_position and self.patroni.failover_priority < st.failover_priority:
+                        # There's a higher priority non-lagging replica
+                        logger.info(
+                            '%s has equally tolerable WAL position and priority %s, while this node has priority %s',
+                            st.member.name,
+                            st.failover_priority,
+                            self.patroni.failover_priority,
+                        )
```

Comment:
I think you need to start from scratch for this block. You reverted some of the changes done in  the https://github.com/zalando/patroni/pull/2804

Reply:
Good catch! Sorry I missed that